### DISCUSSION TITLE
Setup ruff lsp

### DISCUSSION THREAD

**Discussion by amitkot** [LunarVim/LunarVim#3877] [Q&A] [answered]
[ruff](https://github.com/charliermarsh/ruff) is a fast linter for Python. The project provides instructions for installing either using null-ls or directly using ruff-lsp.

### Mason
Although mason [seems to support ruff (and ruff-lsp)](https://github.com/williamboman/mason.nvim/blob/main/PACKAGES.md#ruff), lunarvim seems to be using an older version of mason that does not.
So I installed both ruff and ruff-lsp manually (using brew and pipx, respectively).

### null-ls
The instructions on the ruff project for installing using null-ls are to add:
```lua
local null_ls = require("null-ls")

null_ls.setup({
    sources = {
        null_ls.builtins.formatting.ruff,
        null_ls.builtins.diagnostics.ruff,
    }
})
```

### what eventually worked
These instructions did not work for me with lunarvim, even after I replaced the first line with:
```lua
local null_ls = require("lvim.lsp.null-ls")
```

What eventually worked was installing ruff manually using pipx and adding these lines:
```lua
local configs = require 'lspconfig.configs'
if not configs.ruff_lsp then
  configs.ruff_lsp = {
    default_config = {
      cmd = { 'ruff-lsp' },
      filetypes = { 'python' },
      root_dir = require('lspconfig').util.find_git_ancestor,
      init_options = {
        settings = {
          args = {}
        }
      }
    }
  }
end
require('lspconfig').ruff_lsp.setup {
  on_attach = on_attach,
}
```

Is there a better way to install it?

**Comment by LostInTheLogs** [comment, marked as answer]
If you're not on master then no, we pin our plugins and nullls from 1.2 doesn't support ruff

**Reply by amitkot** [reply]
Can I use specify different plugin versions for specific ones, in my setup?

**Reply by LostInTheLogs** [reply]
Impossible

**Reply by amitkot** [reply]
Okay, thanks for the info. When is the next LunarVim version update expected to be released?

**Reply by LostInTheLogs** [reply]
together with next neovim version
